### PR TITLE
Make passing arguments by value into references

### LINKED ISSUE #3074: Almost all of json parsing is done by value
The issue with passing objects by value instead of reference is pretty widespread in the library. Let's look into the issue more and try to pass as much of stuff (probably not just JSON) as reference as possible.

### PR DESCRIPTION
**Pull request by MartinBriza**
### 📒 Description
Passing compound variables (classes) by value makes them actually copy the passed data instead of just passing a simple reference/pointer (without copying any data).
This PR changes most of value passing in the library into passing by reference.

### 🕶️ Types of changes
- **Bug fix** (non-breaking change which fixes an issue)

### 👫 Relationships
Closes #3074 #3012 

### 🔎 Review hints
Check if general functionality is preserved. This PR should also be reviewed by thorough code reading.

### CODE REVIEW

**Comment by MartinBriza** [comment]
The concept of mutable and immutable variables doesn't really work that way in C++.
By passing the strings by value, you only increase the performance cost of function calls. There is no benefit regarding thread safety - the program has to access the memory of the original string at some point at least once. And yes, there's the chance some other thread will modify the memory location of the string but this is not something passing by value would solve - that's what mutexes and other synchronization primitives are for.

**Comment by NghiaTranUIT** [comment]
@MartinBriza Can you review this branch from master? Your master is pretty old and it has some macOS-related bug, which fixed in last master. Then I could continue checking the logic in macOS side 👍

**Comment by MartinBriza** [comment]
Huh? This branch still merges against `master` just fine. You can just rebase it for merging.

File: src/formatter.cc [hunk 1]
```diff
@@ -28,15 +28,15 @@
 
 namespace toggl {
 
-const std::string Format::Classic = std::string("classic");
-const std::string Format::Improved = std::string("improved");
-const std::string Format::Decimal = std::string("decimal");
+const std::string &Format::Classic = std::string("classic");
+const std::string &Format::Improved = std::string("improved");
+const std::string &Format::Decimal = std::string("decimal");
```
Comment (skel35) [inline src/formatter.cc]: does it make sense to use references here?

Comment (MartinBriza) [reply src/formatter.cc]: Nice catch! While not breaking anything, it doesn't really make sense, you're right. I'll get rid of it.

File: src/websocket_client.cc [hunk 2]
```diff
@@ -195,14 +195,14 @@ error WebSocketClient::receiveWebSocketMessage(std::string *message) {
...
-    } catch(const std::string& ex) {
+    } catch(const std::string & ex) {
         return error(ex);
     }
     *message = json;
     return noError;
 }
 
-const std::string kPong("{\"type\": \"pong\"}");
+const std::string &kPong("{\"type\": \"pong\"}");
```
Comment (skel35) [inline src/websocket_client.cc]: same here

**Comment by IndrekV** [comment]
I'm very curious to see numbers after this is ready and merged. How the performance or memory allocation is compared between the old solution and this update.

**Comment by IndrekV** [comment]
I tested general functionality on Mac and Linux and everything works. We now have historic moment of 3 approvals for a PR. Well done @MartinBriza and the reviewers!

**Pull request merged**